core/iwasm/libraries/lib-socket/test/tcp.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+#include <unistd.h>
+#include <string.h>
+#include <assert.h>
+#ifdef __wasi__
+#include <wasi_socket_ext.h>
+#else
+#include <arpa/inet.h>
+#endif
+#include <pthread.h>
+#define SERVER_MSG "Message from server."
+#define PORT 8989
+pthread_mutex_t mut;
+pthread_cond_t cond;
+int server_init_complete = 0;
+
+struct init_socket {
+    struct sockaddr_in addr;
+    int sock;
+};
+
+struct init_socket
+init_socket_addr()
+{
+    int sock = socket(AF_INET, SOCK_STREAM, 0);
+    assert(sock > 0);
+
+    struct sockaddr_in addr;
+    memset(&addr, 0, sizeof(addr));
+    addr.sin_family = AF_INET;
+    addr.sin_port = PORT;
+    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    return (struct init_socket){ addr, sock };
+}
+
+void *
+server()
+{
+    pthread_mutex_lock(&mut);
```

Comment:
we probably only need to lock the mutex when updating `server_init_complete` variable, i.e. in L53